### PR TITLE
Move charge related code to separate try-catch to prevent renewal failure

### PR DESCRIPTION
**Pull request by Mayisha**
Reported in 8657329-zd-a8c 
Discussion p1726467962851179-slack-C7U3Y3VMY

In [process_subscription_payment](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L232C18-L232C46) function under one try-catch block, we make 2 API calls during the payment flow. 
1. The first API call is to [create and confirm the intent](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L312) which mainly processes the subscription payment. In this step, the card is already charged if the request is successful.
2. We make an API call to get the `charge` object to find the transaction ID. ([here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L373) and [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L403))

The merchant reported that their customers are charged multiple times for the same subscription renewal. They have the retry mechanism enabled for failed payments.
We have noticed from their logs that they encountered some timeouts during requests to Stripe. In the above function, when they encountered the timeouts in step 2, it marked their order as failed. As a result the retry mechanism was triggered and customer was charged again. 

## Changes proposed in this Pull Request:
- Moved the API requests under two separate try-catch blocks. 
- If the first request (create and confirm intent) fails, the card won't be charged. In the catch block of this section, we change the order status to `failed`.
- In the catch block of second API request (get charge), we only log the error but won't change the order status to failed as the card should be already charged.

## Testing instructions

- Checkout to `develop` branch.
- From **WooCommerce > Settings > Subscriptions**, enable `Retry Failed Payments`.
- As a shopper, purchase a subscription.
- In [this line](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L401) add the following code to intentionally fail the get `charges` request.

```
// tweak charge id to fail intentionally
if ( ! empty( $response->latest_charge ) ) {
    $response->latest_charge = $response->latest_charge . 'm';
}
```
- As an admin, go to the subscription edit page and renew the subscription from `Order actions` dropdown.
- Go to the renewal order edit page.
- Notice that the order has failed but in the order notes you will see a payment intent was created.
- Go to your Stripe dashboard and you will find the corresponding transaction.
- Now checkout to this branch.
- Follow the above steps to reproduce the issue.
- This time the order should not fail but you should see the `charges` related error in your logs.

### CODE REVIEW

File: includes/compat/trait-wc-stripe-subscriptions.php [hunk 1]
```diff
@@ -363,6 +363,21 @@ public function process_subscription_payment( $amount, $renewal_order, $retry =
...
+			WC_Stripe_Logger::log( 'Error: ' . $e->getMessage() );
+
+			do_action( 'wc_gateway_stripe_process_payment_error', $e, $renewal_order );
+
+			/* translators: error message */
+			$renewal_order->update_status( 'failed' );
+		}
+
+		try {
+
```
Comment (Mayisha) [inline includes/compat/trait-wc-stripe-subscriptions.php:419]: The diff makes it look like some lines are added or removed. But the only change I made here is, I have moved this [block](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L366-L405) to another try-catch below, where catch block logs the error only.

File: includes/compat/trait-wc-stripe-subscriptions.php [hunk 2]
```diff
@@ -363,6 +363,21 @@ public function process_subscription_payment( $amount, $renewal_order, $retry =
...
 			// Either the charge was successfully captured, or it requires further authentication.
 			if ( $is_authentication_required ) {
```
Comment (mattallan) [inline includes/compat/trait-wc-stripe-subscriptions.php:421]: If an exception is thrown in the top half, variables like `$response` and `$is_authentication_required` won't exist.

Do we need to add a `return;` in the catch statement above, or is it intended to still execute the charge requests even if the payment fails?

**Review by mattallan — CHANGES_REQUESTED** [review]
Hey @Mayisha thanks for the details description and nice approach to this issue!

After performing code review I have a quick question, should we be returning at the end of the first `catch` if the payment fails?

File: includes/compat/trait-wc-stripe-subscriptions.php (hunk 2 above)
Comment (Mayisha) [reply includes/compat/trait-wc-stripe-subscriptions.php:421]: Good catch Matt. I have added `return` in the first catch [92465a1](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3444/commits/92465a1de8e512e5e135521a1247b4fe6a1caed4).

**Review by mattallan — APPROVED** [review]
Thanks @Mayisha for the latest commits!

I've tested these changes and confirm the order is not marked as failed if the request to get the latest charge and process response fails. 💯 The order was left pending payment and then later was automatically marked as processing once the intent success webhook was sent (i'm assuming this is correct).

**Pull request merged**